ledger/participant-integration-api/src/main/scala/platform/usermanagement/PersistentUserManagementStore.scala

```diff
@@ -0,0 +1,259 @@
+                grantedAt = epochMicroSeconds(),
+              )(connection)
+            } else {
+              false
+            }
+          }
+          if (
+            backend.countUserRights(user.internalId)(connection) > maxNumberOfUserRightsPerUserLimit
+          ) {
+            throw TooManyUserRightsRuntimeException(user.domainUser.id)
+          } else {
+            addedRights
+          }
+        }
+      }
+    }.map(tapSuccess { grantedRights =>
+      foreachEachUserRight(grantedRights) { suffix =>
+        logger.info(s"Granted user rights to user id: ${id}. Granted right $suffix")
+      }
+    })(scala.concurrent.ExecutionContext.parasitic)
+  }
+
+  override def revokeRights(
+      id: UserId,
+      rights: Set[domain.UserRight],
+  ): Future[Result[Set[domain.UserRight]]] = {
+    inTransaction { implicit connection =>
+      withUser(id = id) { user =>
+        val revokedRights = rights.filter { right =>
+          if (backend.userRightExists(internalId = user.internalId, right = right)(connection)) {
+            backend.deleteUserRight(internalId = user.internalId, right = right)(connection)
+          } else {
+            false
+          }
+        }
+        revokedRights
+      }
+    }.map(tapSuccess { revokedRights =>
+      foreachEachUserRight(revokedRights) { suffix =>
+        logger.info(s"Revoked user rights from user id: ${id}. Revoked right $suffix")
+      }
+    })(scala.concurrent.ExecutionContext.parasitic)
+
+  }
+
+  override def listUserRights(id: UserId): Future[Result[Set[domain.UserRight]]] = {
+    inTransaction { implicit connection =>
+      withUser(id = id) { user =>
+        backend.getUserRights(internalId = user.internalId)(connection)
+      }
+    }
+  }
+
+  override def listUsers(): Future[Result[Users]] = {
+    inTransaction { connection =>
+      Right(backend.getUsers()(connection))
+    }
+  }
+
+  private def inTransaction[T](thunk: Connection => Result[T]): Future[Result[T]] = {
+    // TODO participant user management: use finer-grained metrics
+    dbDispatcher
+      .executeSql(metrics.daml.userManagement.allUserManagement)(thunk)
+      .recover { case TooManyUserRightsRuntimeException(userId) =>
+        Left(TooManyUserRights(userId))
+      }(ExecutionContext.parasitic)
+  }
+
+  private def withUser[T](
+      id: Ref.UserId
+  )(f: UserManagementStorageBackend.DbUser => T)(implicit connection: Connection): Result[T] = {
+    backend.getUser(id = id)(connection) match {
+      case Some(user) => Right(f(user))
+      case None => Left(UserNotFound(userId = id))
+    }
+  }
+
+  private def withoutUser[T](
+      id: Ref.UserId
+  )(t: => T)(implicit connection: Connection): Result[T] = {
+    backend.getUser(id = id)(connection) match {
+      case Some(user) => Left(UserExists(userId = user.domainUser.id))
+      case None => Right(t)
+    }
+  }
+
+  private def epochMicroSeconds(): Long = Instant.now.getEpochSecond * 1000 * 1000
```

Comment:
This rounds all timestamps to the second. Please extract the microseconds as well.

Reply:
Handling micros from epoch on DB side